src/wasm-lib/kcl/src/parser/parser_impl.rs

```diff
@@ -0,0 +1,2045 @@
+                }))
+            }
+        };
+        Ok(op)
+    })
+    .context(expected("a binary operator (like + or *)"))
+    .parse_next(i)
+}
+
+/// Parse a KCL operand that can be used with an operator.
+fn operand(i: TokenSlice) -> PResult<BinaryPart> {
+    const TODO_783: &str = "found a value, but this kind of value cannot be used as the operand to an operator yet (see https://github.com/KittyCAD/modeling-app/issues/783)";
+    let op = possible_operands
+        .try_map(|part| {
+            let source_ranges = vec![SourceRange([part.start(), part.end()])];
+            let val = match part {
+                // TODO: these should be valid operands eventually,
+                // users should be able to run "let x = f() + g()"
+                // see https://github.com/KittyCAD/modeling-app/issues/783
+                Value::FunctionExpression(_)
+                | Value::PipeExpression(_)
+                | Value::PipeSubstitution(_)
+                | Value::ArrayExpression(_)
+                | Value::ObjectExpression(_) => {
+                    return Err(KclError::Syntax(KclErrorDetails {
+                        source_ranges,
+                        message: TODO_783.to_owned(),
+                    }))
+                }
+                Value::UnaryExpression(x) => BinaryPart::UnaryExpression(x),
+                Value::Literal(x) => BinaryPart::Literal(x),
+                Value::Identifier(x) => BinaryPart::Identifier(x),
+                Value::BinaryExpression(x) => BinaryPart::BinaryExpression(x),
+                Value::CallExpression(x) => BinaryPart::CallExpression(x),
+                Value::MemberExpression(x) => BinaryPart::MemberExpression(x),
+            };
+            Ok(val)
+        })
+        .context(expected("an operand (a value which can be used with an operator)"))
+        .parse_next(i)?;
+    Ok(op)
+}
+
+impl TokenType {
+    fn parse_from(self, i: TokenSlice) -> PResult<Token> {
+        any.try_map(|token: Token| {
+            if token.token_type == self {
+                Ok(token)
+            } else {
+                Err(KclError::Syntax(KclErrorDetails {
+                    source_ranges: token.as_source_ranges(),
+                    message: format!(
+                        "expected {self} but found {} which is a {}",
+                        token.value.as_str(),
+                        token.token_type
+                    ),
+                }))
+            }
+        })
+        .parse_next(i)
+    }
+}
+
+/// Parse some whitespace (i.e. at least one whitespace token)
+fn whitespace(i: TokenSlice) -> PResult<Vec<Token>> {
+    repeat(
+        1..,
+        any.try_map(|token: Token| {
+            if token.token_type == TokenType::Whitespace {
+                Ok(token)
+            } else {
+                Err(KclError::Syntax(KclErrorDetails {
+                    source_ranges: token.as_source_ranges(),
```

Comment:
Is there a reason you aren't using `take_while`?  Not working with custom tokens?

Reply:
I didn't think about take_while. Now that I'm trying it, you're correct it doesn't work with custom tokens -- I get "the trait `AsChar` is not implemented for `token::Token`". I'll file an issue.

Reply:
Filed https://github.com/winnow-rs/winnow/issues/346